text/0000-router-helpers.md

```diff
@@ -0,0 +1,442 @@
+
+### CSS Class Magic
+
+`{{link-to}}` adds some convienent, yet not obvious, classes to the element. These classes are:
+
+- `active`: applied to any `{{link-to}}` that is on the "active" path
+- `disabled`: applied depending on the evaluation of `disabled=someBool`
+- `loading`: applied if one or more of the models passed are `undefined`
+- `ember-transitioning-in`: applied to links that are about to be `active`
+- `ember-transitioning-out`: applied to links that are about to be deactivated
+
+The issue with these class names is that they are not declared anywhere in your templated and are provided by the `LinkComponent` as `classNameBindings`. This effectively creates a set of reserved class names that are highly prone to colissions in your typical application.
+
+Furthermore, addons like [ember-cli-active-link-wrapper](https://github.com/alexspeller/ember-cli-active-link-wrapper) and [ember-bootstarp-nav-link](https://github.com/zoltan-nz/ember-bootstrap-nav-link) do a ton of work arounds to get things like the `.active` class to show up on wrapping elements instead of the element directly. This is a great example that shows we are missing some primitives.
+
+### Default Query Param Serialization
+
+Lastly, `{{link-to}}` has very strange behavior when it comes to serializing query params. On a controller you declare the query params for a specific route. These query params can have defaults for them. For example if you have a controller that looks like:
+
+```js
+// app/controllers/profile.js
+import Controller from '@ember/controller';
+
+export default Controller.extend({
+  queryParams: ['someBool'],
+  someBool: true,
+})
+```
+
+and you to link to it like this:
+
+```hbs
+{{#link-to 'profile'}}Profile{{/link-to}}
+```
+
+In the DOM you will have an `href` on the anchor that gets serializes as:
+
+```html
+<a href="/profile?someBool=true" class="active ember-view">Profile</a>
+```
+
+Looking at a template you would have no idea that rendering the `{{link-to}}` would result in the query params being serialized. From an implementation point of view, this is problematic as we are forced to `lookup` the `Route` and the associated `Controller` to grab the query params. This can add a non-trivial amount of overhead during rendering, especially if you have many `{{link-to}}`s on a route that link many different parts of your application. As a side-note, this is one of the things  that needs to be dealt with if we are ever to kill controllers.
```

Comment:
Does it have to be "kill controllers"? :X

```suggestion
Looking at a template you would have no idea that rendering the `{{link-to}}` would result in the query params being serialized. From an implementation point of view, this is problematic as we are forced to `lookup` the `Route` and the associated `Controller` to grab the query params. This can add a non-trivial amount of overhead during rendering, especially if you have many `{{link-to}}`s on a route that link many different parts of your application. As a side-note, this is one of the things that needs to be dealt with if we are ever to kill controllers.
```